### DISCUSSION TITLE
Can I disable the gatsby-react-router-scroll for hash fragments?

### DISCUSSION THREAD

**Discussion by OssiPesonen** [gatsbyjs/gatsby#26437] [Help]
## Summary

I recently wanted to try out Contentful with Gatsby and Netlify. It's what Contentful proposed. After this I went and implemented user login via Contentful which returns an access token in the URI as hash fragment.

> You can specify a custom redirect URI that will receive the access token as part of the URI's hash fragment.

ie: `page#access_token=<ACCESS_TOKEN>&token_type=Bearer`

But it appears Gatsby's `gatsby-react-router-scroll` attempts to read this as an identifier and scroll to that position, causing the page to crash. I would like to disable the feature or perhaps ask you to fix this behavior.

## Relevant information

Contentful redirects the user with the access token back to a `redirect_uri` which I've set to `/code` (src/pages/code.js) and this page crashes by default.

```
SyntaxError: Document.querySelector: '#access_token=<ACCESS_TOKEN>&token_type=Bearer' is not a valid selector
ScrollHandler/_this.scrollToHash
http://localhost:8000/commons.js:19751:27

  19748 | };
  19749 | 
  19750 | _this.scrollToHash = function (hash, prevProps) {
> 19751 |   var node = document.querySelector(hash);
        |                       ^  19752 | 
  19753 |   if (node && _this.shouldUpdateScroll(prevProps, _this.props)) {
  19754 |     node.scrollIntoView();

componentDidMount
G:/Dropbox/Dev/www/starter-gatsby-blog/node_modules/gatsby-react-router-scroll/scroll-handler.js:89

  86 |   if (scrollPosition) {
  87 |     this.windowScroll(scrollPosition, undefined);
  88 |   } else if (hash) {
> 89 |     this.scrollToHash(decodeURI(hash), undefined);
  90 |   }
  91 | };
  92 | 

`````

I've tried adding `gatsby-browser.js` and overwriting various functions such as `shouldUpdateScroll` and `scrollToHash` but these have no effect.

### Environment (if relevant)

```
  System:
    OS: Windows 10 10.0.18362
    CPU: (12) x64 AMD Ryzen 5 3600 6-Core Processor              
  Binaries:
    Node: 12.16.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.4 - G:\Dropbox\Dev\www\starter-gatsby-blog\node_modules\.bin\yarn.CMD
    npm: 6.13.4 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 3.6.5 - C:\Users\Admin\AppData\Local\Programs\Python\Python36-32\python.EXE
  Browsers:
    Edge: 44.18362.449.0
  npmPackages:
    gatsby: ^2.23.9 => 2.23.9 
    gatsby-image: ^2.4.8 => 2.4.8 
    gatsby-plugin-react-helmet: ^3.3.5 => 3.3.5 
    gatsby-plugin-use-query-params: ^1.0.1 => 1.0.1 
    gatsby-source-contentful: ^2.3.17 => 2.3.17 
    gatsby-transformer-remark: ^2.8.18 => 2.8.18 
    gatsby-transformer-sharp: ^2.5.6 => 2.5.6 
```

### File contents (if changed)

`gatsby-config.js`: N/A 
`package.json`: N/A 
`gatsby-node.js`: N/A 
`gatsby-browser.js`: N/A 
`gatsby-ssr.js`: N/A

**Comment by jfgilmore** [comment]
I have a possibly related issue to this in that the `shouldUpdateScroll` API in `gatsby-browser` appears not to be triggered by any link containing a `hash` value in the location. The usual workaround for your problem @OssiPesonen would be to use `shouldUpdateScroll` to intercept any value of `routerProps.location.hash` of interest to you and return `false` so that the scroll position is not handled by Gatsby. However if it is never being called when the location props contain a `hash` as I suspect, then you can never intercept it.

In my case I wish to make sure my DOM elements are ready before it tries to scroll. However I have discovered that the `node` value in `scrollToHash` always returns `null` and so `node.scrollIntoView()` is skipped. As it turns out so is `this.shouldUpdateScroll(prevProps, this.props)` wherein lies the problem and the solution.

I am going to submit a PR with my proposed patch, I have tested this in my own source code and it appears to fix both of these problems.
